pkg/genericapiserver/enablement_config.go

```diff
@@ -0,0 +1,159 @@
+	// DisabledResources trumps EnabledResources
+	DisabledResources sets.String
+
+	// EnabledResources lists the resources that should be enabled by default.  This is a little
+	// unusual, but we need it for compatibility with old code for now.  An empty set means
+	// enable all, a non-empty set means that all other resources are disabled.
+	EnabledResources sets.String
+}
+
+var _ APIResourceEnablement = &APIResourceEnablementConfig{}
+
+// AnyVersion is used during resource enablement checks to see if any version of the resource is enabled
+const AnyVersion = "."
+
+type APIResourceEnablementConfig struct {
+	GroupVersionDisablers map[unversioned.GroupVersion]*APIGroupVersionEnablementConfig
+}
+
+func NewAPIResourceEnablementConfig() *APIResourceEnablementConfig {
+	return &APIResourceEnablementConfig{GroupVersionDisablers: map[unversioned.GroupVersion]*APIGroupVersionEnablementConfig{}}
+}
+
+func NewAPIGroupVersionEnablementConfig() *APIGroupVersionEnablementConfig {
+	return &APIGroupVersionEnablementConfig{Enable: true, DisabledResources: sets.String{}, EnabledResources: sets.String{}}
+}
+
+func (o *APIResourceEnablementConfig) DisableVersions(versions ...unversioned.GroupVersion) {
+	for _, version := range versions {
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].Enable = false
+	}
+}
+
+func (o *APIResourceEnablementConfig) EnableVersions(versions ...unversioned.GroupVersion) {
+	for _, version := range versions {
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].Enable = true
+	}
+}
+
+func (o *APIResourceEnablementConfig) DisableResources(resources ...unversioned.GroupVersionResource) {
+	for _, resource := range resources {
+		version := resource.GroupVersion()
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].DisabledResources.Insert(resource.Resource)
+	}
+}
+
+func (o *APIResourceEnablementConfig) EnableResources(resources ...unversioned.GroupVersionResource) {
+	for _, resource := range resources {
+		version := resource.GroupVersion()
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].EnabledResources.Insert(resource.Resource)
+		o.GroupVersionDisablers[version].DisabledResources.Delete(resource.Resource)
+	}
+}
+
+func (o *APIResourceEnablementConfig) ResourceEnabled(resource unversioned.GroupVersionResource) bool {
+	if resource.Version == AnyVersion {
```

Comment:
Why have this special flag? Why not make the AnyVersionOfResourceEnabled function public?